### ISSUE TITLE
Applying Mosaic causes keypoints outside the image to move to the border.

### ISSUE THREAD

**Issue by HeadTriXz** [ultralytics/ultralytics#19920]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training a YOLO Pose model with mosaic augmentation enabled, the keypoints that would otherwise be outside the image, are moved to the border. This seems to happen when the random center causes some images to be cut off:
https://github.com/ultralytics/ultralytics/blob/5c39f5ed8502e991dcbf6ba5f8f750206563e582/ultralytics/data/augment.py#L684

I assume this happens due to clipping the instances:
https://github.com/ultralytics/ultralytics/blob/5c39f5ed8502e991dcbf6ba5f8f750206563e582/ultralytics/data/augment.py#L859

![Image](https://github.com/user-attachments/assets/cfac345f-f308-44a2-b9d9-adb56edea784)

### Environment

Ultralytics 8.3.97 🚀 Python-3.11.11 torch-2.6.0+cu126 CUDA:0 (NVIDIA GeForce RTX 4070 Ti SUPER, 16376MiB)
Setup complete ✅ (24 CPUs, 31.7 GB RAM, 256.1/951.6 GB disk)

OS                  Windows-10-10.0.26100-SP0
Environment         Windows
Python              3.11.11
Install             git
Path                .venv\Lib\site-packages\ultralytics
RAM                 31.74 GB
Disk                256.1/951.6 GB
CPU                 AMD Ryzen 9 7900 12-Core Processor
CPU count           24
GPU                 NVIDIA GeForce RTX 4070 Ti SUPER, 16376MiB
GPU count           1
CUDA                12.6

numpy               ✅ 2.1.1<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0+cu126>=1.8.0
torch               ✅ 2.6.0+cu126!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0+cu126>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

```py
import cv2
import yaml
from ultralytics.data import YOLODataset
from ultralytics.data.augment import Mosaic


def load_yaml(path: str) -> dict:
    with open(path, "r") as f:
        return yaml.safe_load(f)


def main() -> None:
    dataset = YOLODataset(
        data=load_yaml("./data/dataset/dataset.yaml"),
        img_path="./data/dataset/images/train",
        imgsz=1280,
        task="pose",
    )

    transform = Mosaic(
        dataset=dataset,
        imgsz=1280,
    )

    while True:
        data = dataset.get_image_and_label(0)
        aug = transform(data)

        image = aug["img"]
        for keypoints in aug["instances"].keypoints:
            for x, y, v in keypoints:
                if v != 2:
                    continue

                image = cv2.circle(image, (int(x), int(y)), 5, (255, 0, 0), -1)

        cv2.imshow("Image", image)
        if cv2.waitKey(0) & 0xFF == ord("q"):
            break


if __name__ == "__main__":
    main()
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
@Laughing-q I guess the visibility for these keypoints need to be set to 0 when clipping

**Comment by HeadTriXz** [comment]
> [@Laughing-q](https://github.com/Laughing-q) I guess the visibility for these keypoints need to be set to 0 when clipping

See #19921

**Comment by UltralyticsAssistant** [comment]
The issue may now be resolved with the merge of the linked PR [#19921](https://github.com/ultralytics/ultralytics/pull/19921), titled "Fix keypoints being clipped rather than hidden." Huge thanks to @HeadTriXz for authoring the fix and @glenn-jocher for contributing to the improvements! 🙌  

### Key Changes  
This PR enhances the handling of keypoints by ensuring that keypoints outside the image boundaries are marked as invisible instead of clipping their coordinates. This change improves accuracy and reliability for keypoint detection, especially in edge cases, leading to better data integrity. 🚀  

### How to Test  
You can test the changes by updating your environment with the latest updates from the main branch:  
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```  
Alternatively, you can wait for the next official PyPI release, where this fix will be included.  

### Feedback Request  
Please let us know if these changes resolve the issue for you. Your feedback is invaluable in ensuring everything works as expected!  

Thank you for reporting this issue and helping us improve. If the problem persists or you encounter anything else, feel free to share additional details. 🙏

**Comment by glenn-jocher** [comment]
Thank you for reporting this issue and submitting the fix in PR #19921! The problem occurred because keypoints outside image boundaries were being clipped to the border rather than being marked as invisible. This created incorrect training data where keypoints that should be out of view appeared at the image edges.

Looking at the code you referenced, the issue was in the `clip` method in `instance.py` where keypoints were having their coordinates clipped but their visibility wasn't being properly updated. The fix properly sets the visibility to 0 for keypoints that fall outside image boundaries.

This is an excellent contribution that improves the accuracy of pose estimation models. If you notice any other issues with the keypoint handling after applying the fix, please let us know.
